async-rwlock/src/lib.rs

```diff
@@ -636,6 +657,80 @@ impl<T: ?Sized> Drop for RwLockWriteGuard<'_, T> {
     }
 }
 
+/// A guard that releases the write lock when dropped.
+pub struct RwLockWriteGuard<'a, T: ?Sized> {
+    writer: RwLockWriteGuardInner<'a, T>,
+    reserved: MutexGuard<'a, ()>,
+}
+
+unsafe impl<T: Send + ?Sized> Send for RwLockWriteGuard<'_, T> {}
+unsafe impl<T: Sync + ?Sized> Sync for RwLockWriteGuard<'_, T> {}
+
+impl<'a, T: ?Sized> RwLockWriteGuard<'a, T> {
+    /// Converts this guard into a reader guard.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # futures_lite::future::block_on(async {
+    /// use async_rwlock::{RwLock, RwLockWriteGuard};
+    ///
+    /// let lock = RwLock::new(1);
+    ///
+    /// let mut writer = lock.write().await;
+    /// *writer += 1;
+    ///
+    /// assert!(lock.try_read().is_none());
+    ///
+    /// let reader = RwLockWriteGuard::downgrade(writer);
+    /// assert_eq!(*reader, 2);
+    /// 
+    /// assert!(lock.try_read().is_some());
+    /// # })
+    /// ```
+    #[inline]
+    pub fn downgrade(guard: Self) -> RwLockReadGuard<'a, T> {
+        guard.writer.0.state.fetch_add(ONE_READER, Ordering::SeqCst);
```

Comment:
Shouldn't this add `ONE_READER - WRITER_BIT`?

Reply:
> Ah, now I see that you just copied parking_lot's API :) In that case, it's probably a safe bet to just copy whatever was done there.

No I didn't just copy the parking_lot api for the copied one. I took it back because it seems more relevant and easier to use. If I just need to downgrade a writer to a reader, it doesn't make sense to force me to downgrade twice: it's harder to use, it forces me to import an intermediate type that I don't need, and it's less efficient :)

Reply:
Indeed